## Importing the necessary libraries

In [7]:
import tensorflow as tf
import numpy as np
import pandas as pd

## Loading the Dataset

In [8]:
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()

# Normalizing the dataset
x_train = x_train/255.0
x_test = x_test/255.0


## Building the Neural Network

In [13]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model.add(tf.keras.layers.Dense(10, activation = tf.nn.softmax))  # we can also write this as activation = 'softmax'
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
# NN has 2 hidden layers   

## Defining a Function to train the model
##  

Here, we have defined a function, which takes the data and model as inputs and returns the loss and accuracy as the output.

The model consists of 3 main steps :
- Defining the callback class and instantiating it to stop training and prevent overfitting
- Fitting the model on the training set
- Evaluating the model's performance on the test set and calculating accuracy

In [10]:
def train_model(x_train,y_train,x_test,y_test,model):
    '''
    function which trains the model and returns the number of epochs as welll as accuracy 
    '''
    class Call_back(tf.keras.callbacks.Callback):
        def on_epoch_end(self,epoch,logs={}):
            if(logs.get('accuracy')>0.99):
                print('Desired Accuracy Reached.Stopping the Training.')
                self.model.stop_training = True
    
    callback = Call_back()
                

    
    # model fitting
    history = model.fit(x_train,y_train,callbacks = [callback],epochs =10)
    metrics = model.evaluate(x_test,y_test)
    return metrics

#### Now, we call the train model function...

In [11]:
train_model(x_train,y_train,x_test,y_test,model)

Train on 60000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 156us/sample - loss: 0.2288 - accuracy: 0.9328
Epoch 2/10
60000/60000 [==============================] - 10s 160us/sample - loss: 0.0967 - accuracy: 0.9699
Epoch 3/10
60000/60000 [==============================] - 10s 161us/sample - loss: 0.0692 - accuracy: 0.9781
Epoch 4/10
60000/60000 [==============================] - 8s 140us/sample - loss: 0.0518 - accuracy: 0.9830
Epoch 5/10
60000/60000 [==============================] - 9s 156us/sample - loss: 0.0403 - accuracy: 0.9870
Epoch 6/10
60000/60000 [==============================] - 8s 137us/sample - loss: 0.0330 - accuracy: 0.9891
Epoch 7/10
10000/10000 [==============================] - 1s 84us/sample - loss: 0.0844 - accuracy: 0.9780


[0.08441498854700621, 0.978]

### As we can see we have a loss of 0.08 and an accuracy of 97% on the validation set 